avocado/core/runners/utils/messages.py

```diff
@@ -0,0 +1,198 @@
+    """Prepare a message dict with necessary information for specific type.
+
+    :param msg: message data. If the message is str, it will be encoded
+                with utf-8.
+    :type msg: str, bytes
+    :param message_type: specific type of message
+    :type message_type: str
+    :return: message dict which can be send to avocado server
+    :rtype: dict
+    """
+    message = {'type': message_type, 'log': msg}
+    if type(msg) is not bytes:
+        msg = msg.encode('utf-8')
+        message.update({'log': msg, 'encoding': 'utf-8'})
+    return _prepare_message('running', message)
+
+
+def get_started_message():
+    """Creates the started message.
+
+    :return: started message
+    :rtype: dict
+    """
+    return _prepare_message('started')
+
+
+def get_finished_message(result, fail_reason=None, returncode=None):
+    """Creates finished message with  all necessary information.
+
+    :param result: test result
+    :type result values for the statuses defined in
+                 :class: avocado.core.teststatus.STATUSES
+    :param fail_reason: parameter for brief specification, of the failed
+                        result.
+    :type fail_reason: str
+    :param returncode: exit status of runner
+    :return: finished message
+    :rtype: dict
+    """
+    message = {'result': result}
+    if fail_reason is not None:
+        message['fail_reason'] = fail_reason
+    if returncode is not None:
+        message['returncode'] = returncode
+    return _prepare_message('finished', message)
+
+
+def get_running_message():
+    """Creates running message without any additional info.
+
+    :return: running message
+    :rtype: dict
+    """
+    return _prepare_message('running')
+
+
+def get_log_message(message):
+    """Creates log message with  all necessary information.
+
+    :param message: log to be sent. The str will be encoded with utf-8.
+    :type message: str, bytes
+    :return: log message
+    :rtype: dict
+    """
+    return _get_running_message(message, 'log')
+
+
+def get_stdout_message(message):
+    """Creates stdout message with  all necessary information.
+
+    :param message: data to be sent. The str will be encoded with utf-8.
+    :type message: str, bytes
+    :return: stdout message
+    :rtype: dict
+    """
+    return _get_running_message(message, 'stdout')
+
+
+def get_stderr_message(message):
+    """Creates stderr message with  all necessary information.
+
+    :param message: data to be sent. The str will be encoded with utf-8.
+    :type message: str, bytes
+    :return: stderr message
+    :rtype: dict
+    """
+    return _get_running_message(message, 'stderr')
+
+
+def get_whiteboard_message(message):
+    """Creates whiteboard message with  all necessary information.
+
+    :param message: data to be sent. The str will be encoded with utf-8.
+    :type message: str, bytes
+    :return: whiteboard message
+    :rtype: dict
+    """
+    return _get_running_message(message, 'whiteboard')
```

Comment:
I have a feeling that we could implement this using a OOP approach, some thing like this:

```python
class GenericMessage:
    message_type = None

    @classmethod
    def _prepare_message(self, additional_info=None):
        """Prepare a message dict with some basic information.
        This will add the keyword 'status' and 'time' to all messages.
        :param status_type: The type of event ('started', 'running',
                             'finished')
        :param: addional_info: Any additional information that you
                               would like to add to the message.
        :type additional_info: dict
        :rtype: dict
        """
        status = {}
        if isinstance(additional_info, dict):
            status = additional_info
        status.update({'status': self.message_type,
                       'time': time.monotonic()})
        return status

    @classmethod
    def get(cls):
        return cls._prepare_message()


class StartedMessage(GenericMessage):
    message_type = 'started'

class RunningMessage(GenericMessage):
    message_type = 'running'

    def log(cls):
        pass

    def stdout(cls):
        pass

    ... 
```

The code is not tested, just to give you some ideas. What do you think?

Reply:
Yes, great idea, but I will have to do more tests based on the concern which @clebergnu mentioned about podman spawner.